analyzers/src/SonarAnalyzer.Common/Rules/UseTestableTimeProviderBase.cs

```diff
@@ -0,0 +1,62 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UseTestableTimeProviderBase<TSyntaxKind> : SonarDiagnosticAnalyzer
+         where TSyntaxKind : struct
+    {
+        protected const string DiagnosticId = "ToBedesided";
+        protected const string MessageFormat = "Use a testable (date) time provider instead.";
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+        protected UseTestableTimeProviderBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected sealed override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+            {
+                if (c.SemanticModel.GetSymbolInfo(c.Node).Symbol is IMethodSymbol identifier
```

Comment:
this will be very expensive in terms of performance - we usually avoid doing symbol retrieval (which triggers a compilation beneath) at the beginning of a rule. 

This rule can easily be syntax-based. You can check the name of the Identifier and MemberAccess (and ignore the edge case of static using), and eventually do a sanity symbol check after verifying it's `DateTime.X`

Reply:
You're right. I think I saw it being done this way elsewhere, but I'll rewrite in such a way that it first checks for the naming only.